examples/platforms/nrf528xx/src/radio.c

```diff
@@ -860,11 +878,38 @@ void nrf_802154_receive_failed(nrf_802154_rx_error_t error)
     setPendingEvent(kPendingEventReceiveFailed);
 }
 
+#if OPENTHREAD_CONFIG_CSL_RECEIVER_ENABLE
+uint16_t getCslPhase()
+{
+    int32_t diff = sCslSampleTime - otPlatAlarmMicroGetNow();
```

Comment:
You mean wrap the function with `OPENTHREAD_CONFIG_PLATFORM_USEC_TIMER_ENABLE`?

Reply:
I mean timer overflow. When timer reaches 0xffffffff, it will start from 0 again. Do we need some extra handling to guard against this?

Reply:
I implemented it as follow now:
```
uint16_t getCslPhase()
{
    uint32_t curTime       = otPlatAlarmMicroGetNow();
    uint32_t cslPeriodInUs = sCslPeriod * OT_US_PER_TEN_SYMBOLS;
    uint32_t diff = ((sCslSampleTime % cslPeriodInUs) - (curTime % cslPeriodInUs) + cslPeriodInUs) % cslPeriodInUs;

    return (uint16_t)(diff / OT_US_PER_TEN_SYMBOLS);
}
```
The idea is to find the nearest coming point(which has the same position with `sCslSampleTime` in one period). By such handling, no matter which is larger, it could be handled correctly. I think the only thing matters is the two points' relative positions in one period. So even when timer reaches 0xffffffff, the modulo of it is still correct. I would like to have your double check.

Reply:
IIUC, from your code, `sCslSampleTime` is the next target wake up time in us. This is calculated with ```c++
mCslSampleTime += mCslPeriod * OT_US_PER_TEN_SYMBOLS;
otPlatRadioUpdateCslSampleTime(mCslSampleTime.GetValue());
```
Let's take a negative example, current time is 0xFFFF FFFE, `sCslSampleTime` is target for 0x3(timer wrap around happens). `mCslPeriod` is 1000. Then the diff should be absolute distance between 0x3 and 0xFFFF FFFE, which is 0x5. However, with your formula, it is (294 - 3 + 1000) % 1000 = 291. Which looks not correct to me.

I think the point here is to consider timer as a ring and calculate the absolute diff.
`uint32_t diff = ((uint32_t) (sCslSampleTime - curTime)) % mCslPeriod;`

ref: https://stackoverflow.com/questions/2061245/how-to-subtract-two-unsigned-ints-with-wrap-around-or-overflow

Reply:
I think above is not a final solution to your problem for `getCslPhase()`. You must handle the case when `curTime` is a little bit ahead of `sCslSampleTime` and no wrap around happens. For example, `sCslSampleTime` is 1000 and `curTime` is 1050. The diff should be 950 rather than the result with the formula I provided.

Reply:
> IIUC, from your code, `sCslSampleTime` is the next target wake up time in us. This is calculated with
> 
> ```c++
> mCslSampleTime += mCslPeriod * OT_US_PER_TEN_SYMBOLS;
> otPlatRadioUpdateCslSampleTime(mCslSampleTime.GetValue());
> ```
> 
> Let's take a negative example, current time is 0xFFFF FFFE, `sCslSampleTime` is target for 0x3(timer wrap around happens). `mCslPeriod` is 1000. Then the diff should be absolute distance between 0x3 and 0xFFFF FFFE, which is 0x5. However, with your formula, it is (294 - 3 + 1000) % 1000 = 291. Which looks not correct to me.
> 
> I think the point here is to consider timer as a ring and calculate the absolute diff.
> `uint32_t diff = ((uint32_t) (sCslSampleTime - curTime)) % mCslPeriod;`
> 
> ref: https://stackoverflow.com/questions/2061245/how-to-subtract-two-unsigned-ints-with-wrap-around-or-overflow

@brianljt Thanks for the very careful verification. If we put `sCslSampleTime` as `3` and current time as `0xFFFF FFFE` in my formula:
```
uint32_t diff = ((sCslSampleTime % cslPeriodInUs) - (curTime % cslPeriodInUs) + cslPeriodInUs) % cslPeriodInUs;
```
I think it's: `(3 - 294 + 1000) % 1000 = 709`. The value of `curTime` and `sCslSampleTime` in your calculation seems get inversed. Correct me if I'm wrong.

However I do think there's a problem about the minus between the two unsigned integers. `3 - 294` will become a very large number here. And I have verified it. Thanks for pointing it out!

After we solve this unsigned integer problem, I think this can also work for the case when **curTime is ahead of sSampleTime**. We can try with `sSampleTime` as `1000`, `curTime` as `1050` and `CslPeriod` as `1000`. If we put them into my expression:
```
((1000 % 1000) - (1050 % 1000) + 1000) % 1000 = 950
```
It seems to be right. Thoughts?

Reply:
I think we can change the formula as:
```
uint32_t diff = (((sCslSampleTime % cslPeriodInUs) + cslPeriodInUs) - (curTime % cslPeriodInUs)) % cslPeriodInUs;
```
In this way we can avoid getting negative value in the calculating process.

Reply:
LGTM, I think you can put `+ cslPeriodInUs` ahead to avoid uint32_t to int32_t conversion.

```c++
uint32_t diff = (cslPeriodInUs + (sCslSampleTime % cslPeriodInUs) - (curTime % cslPeriodInUs) ) % cslPeriodInUs;
```

Reply:
Exactly. We have "comment racing condition" in the air.

Reply:
On a second thought, there is another problem. For example, current time 0xFFFF FFFE, expected sample time 0x3. Expected diff is 0x5. However, calculated diff is (3 + 1000 - 294) % 1000 = 709. It is still wrong.

The problem here is, 0xFFFF FFFF cannot be divided by 1000. Which means, for the first round, even the sample time happens at 0, 1000, 2000, 3000... The second round of clock does not exactly happens at 0, 1000, 2000... If both of current time and sample time falls within same round of clock, then everything is fine because this method will have same offset for both modulation. However, thing will go wrong if one clock is in different round of another. Am I right?

Reply:
I think we can use this formula,

```c++
uint32_t diff = (cslPeriodInUs + ((uint32_t) (sCslSampleTime - curTime)
) % cslPeriodInUs;
```

Please help double check.

Reply:
That makes sense. LGTM.